File: web/src/pages/relatorio-chuva-acumulada.ts
// Página Chuva Acumulada (spec 022) — relatório puramente tabular.
//
// Ranking de precipitação acumulada por município na janela, com dias de
// chuva e o maior volume diário (+ data). Filtros: período em dias +
// macrorregião opcional. Deep link ?dias=N&meso=… + botão Compartilhar,
// e paginação client-side (10 + "Ver mais") — spec 024.

import { fmt1, formatarDataISO } from "../format";
import { compartilharWhatsapp, esconderCompartilhar, escreverURL, lerURL } from "../share";
import { renderTable, type RowDef } from "../table";
import { byId, setText, toggle } from "../ui";

interface ChuvaRow {
  city_name: string;
  mesoregion: string | null;
  precip_acumulada: number | null;
  dias_chuva: number;
  maior_dia_mm: number | null;
  maior_dia_data: string | null;
}

const CHUVA_PAGE = 10;
let chuvaMax = CHUVA_PAGE;
let ultimoRows: ChuvaRow[] = [];

function clampDias(raw: string | null): number {
  const n = Number.parseInt(raw ?? "", 10);
  if (Number.isNaN(n)) return 30;
  return Math.min(365, Math.max(7, n));
}

function selecionarSePresente(select: HTMLSelectElement, valor: string | null): void {
  if (valor !== null && [...select.options].some((o) => o.value === valor)) {
    select.value = valor;
  }
}

async function popularMesorregioes(select: HTMLSelectElement): Promise<void> {
  const resposta = await fetch("/api/v1/ref/mesorregioes");
  if (!resposta.ok) return;
  for (const meso of (await resposta.json()) as string[]) {
    const opt = document.createElement("option");
    opt.value = meso;
    opt.textContent = meso;
    select.appendChild(opt);
  }
}

async function carregarCaption(): Promise<void> {
  const resposta = await fetch("/api/v1/ref/daily-meta");
  if (!resposta.ok) return;
  const meta = (await resposta.json()) as { max_date: string | null };
  if (meta.max_date !== null) {
    setText("chuva-caption", `Dados disponíveis até ${formatarDataISO(meta.max_date)}`);
  }
}

// Re-render in-place a partir de `ultimoRows` + `chuvaMax` (sem novo fetch).
function renderPagina(): void {
  const tabela = byId<HTMLTableElement>("tabela-chuva");
  const btnMais = byId<HTMLButtonElement>("btn-mais-chuva");
  if (tabela === null) return;

  const rows: RowDef[] = ultimoRows.map((r) => ({
    cells: [
      r.city_name,
      r.mesoregion,
      fmt1(r.precip_acumulada),
      String(r.dias_chuva),
      fmt1(r.maior_dia_mm),
      formatarDataISO(r.maior_dia_data),
    ],
  }));
  const ok = renderTable(tabela, rows, {
    limit: chuvaMax,
    onEmpty: () => {
      toggle(btnMais, false);
      setText("chuva-msg", "Sem dados para o período selecionado.");
    },
  });
  if (!ok) return;
  toggle(byId("chuva-msg"), false);

  const restantes = Math.max(0, ultimoRows.length - chuvaMax);
  if (btnMais !== null) {
    btnMais.hidden = restantes <= 0;
    btnMais.textContent = `Ver mais (${restantes} restantes)`;
  }
}

async function atualizar(dias: number, meso: string): Promise<void> {
  const tabela = byId<HTMLTableElement>("tabela-chuva");
  if (tabela === null) return;
  chuvaMax = CHUVA_PAGE; // filtro novo → volta ao topo
  const params = new URLSearchParams({ dias: String(dias) });
  if (meso !== "Todas") params.set("meso", meso);
  escreverURL(params);
  esconderCompartilhar();

  const resposta = await fetch(`/api/v1/relatorio-chuva-acumulada/dados?${params.toString()}`);
  if (!resposta.ok) {
    setText("chuva-msg", `Erro ao carregar o relatório (HTTP ${resposta.status})`);
    toggle(tabela, false);
    toggle(byId("btn-mais-chuva"), false);
    return;
  }
  ultimoRows = ((await resposta.json()) as { rows: ChuvaRow[] }).rows;
  renderPagina();

  if (ultimoRows.length > 0) {
    const alvo = meso === "Todas" ? "Santa Catarina" : meso;
    compartilharWhatsapp(`Chuva acumulada — ${alvo}, últimos ${dias} dias.`, params);
  }
}

export function initRelatorioChuvaAcumulada(): void {
  const diasInput = byId<HTMLInputElement>("filtro-dias");
  const mesoSelect = byId<HTMLSelectElement>("filtro-mesorregiao");
  if (diasInput === null || mesoSelect === null) return;

  const url = lerURL();
  const dias = clampDias(url.get("dias") ?? diasInput.value);
  diasInput.value = String(dias);

  void carregarCaption();
  void popularMesorregioes(mesoSelect).then(() => {
    selecionarSePresente(mesoSelect, url.get("meso"));
    void atualizar(clampDias(diasInput.value), mesoSelect.value);
  });

  const rerender = (): void => void atualizar(clampDias(diasInput.value), mesoSelect.value);
  diasInput.addEventListener("change", rerender);
  mesoSelect.addEventListener("change", rerender);
  byId("btn-mais-chuva")?.addEventListener("click", () => {
    chuvaMax += CHUVA_PAGE;
    renderPagina();
  });
}

File: api/app/routers/relatorio_chuva_acumulada.py
"""Endpoint JSON da página Chuva Acumulada (spec 022).

- GET /api/v1/relatorio-chuva-acumulada/dados -> ranking de precipitação
  acumulada por município na janela (todas as cidades da macrorregião, ou
  todas do estado), ordenado do maior para o menor.

Sem linha de total — soma de chuva entre municípios não é grandeza física
útil.
"""

from fastapi import APIRouter, Query

from app.routers.ref import meso_filter
from app.schemas.relatorio_chuva_acumulada import (
    ChuvaAcumuladaRow,
    RelatorioChuvaAcumuladaResponse,
)
from app.utils.bigquery import max_date, query, tbl

router = APIRouter(
    prefix="/relatorio-chuva-acumulada", tags=["relatorio-chuva-acumulada"]
)

_DAILY = "mart_climate__daily_facts"


@router.get("/dados", response_model=RelatorioChuvaAcumuladaResponse)
def get_dados(
    dias: int = Query(30, ge=7, le=365, description="Janela em dias (7–365, default 30)"),
    meso: str | None = Query(None, description="Macrorregião (omitir ou 'Todas' = sem filtro)"),
) -> RelatorioChuvaAcumuladaResponse:
    anchor = max_date(_DAILY)
    if anchor is None:
        return RelatorioChuvaAcumuladaResponse()
    clause, meso_params = meso_filter(meso)

    rows = query(
        f"""
        SELECT
          city_name,
          mesoregion,
          ROUND(SUM(precipitation_mm), 1) AS precip_acumulada,
          COUNTIF(precipitation_mm > 0)   AS dias_chuva,
          ROUND(MAX(precipitation_mm), 1)  AS maior_dia_mm,
          MAX_BY(date, precipitation_mm)   AS maior_dia_data
        FROM {tbl(_DAILY)}
        WHERE date >= DATE_SUB(@max_date, INTERVAL @dias DAY)
          {clause}
        GROUP BY city_name, mesoregion
        ORDER BY precip_acumulada DESC, city_name
        """,
        params={"max_date": anchor, "dias": dias, **meso_params},
    )
    return RelatorioChuvaAcumuladaResponse(
        rows=[ChuvaAcumuladaRow(**r) for r in rows]
    )
